torchvision/models/detection/fcos.py

```diff
@@ -0,0 +1,696 @@
+from ..resnet import resnet50
+from . import _utils as det_utils
+from ._utils import overwrite_eps
+from .anchor_utils import AnchorGenerator
+from .backbone_utils import _resnet_fpn_extractor, _validate_trainable_layers
+from .transform import GeneralizedRCNNTransform
+
+
+__all__ = ["FCOS", "fcos_resnet50_fpn"]
+
+
+class FCOSHead(nn.Module):
+    """
+    A regression and classification head for use in FCOS.
+
+    Args:
+        in_channels (int): number of channels of the input feature
+        num_anchors (int): number of anchors to be predicted
+        num_classes (int): number of classes to be predicted
+        num_convs (Optional[int]): number of conv layer of head. Default: 4.
+    """
+
+    __annotations__ = {
+        "box_coder": det_utils.BoxLinearCoder,
+    }
+
+    def __init__(self, in_channels: int, num_anchors: int, num_classes: int, num_convs: Optional[int] = 4):
+        super().__init__()
+        self.box_coder = det_utils.BoxLinearCoder(normalize_by_size=True)
+        self.classification_head = FCOSClassificationHead(in_channels, num_anchors, num_classes, num_convs)
+        self.regression_head = FCOSRegressionHead(in_channels, num_anchors, num_convs)
+
+    def compute_loss(
+        self,
+        targets: List[Dict[str, Tensor]],
+        head_outputs: Dict[str, Tensor],
+        anchors: List[Tensor],
+        matched_idxs: List[Tensor],
+    ):
+
+        cls_logits = head_outputs["cls_logits"]  # [N, HWA, C]
+        bbox_regression = head_outputs["bbox_regression"]  # [N, HWA, 4]
+        bbox_ctrness = head_outputs["bbox_ctrness"]  # [N, HWA, 1]
+
+        all_gt_classes_targets = []
+        all_gt_boxes_targets = []
+        for targets_per_image, matched_idxs_per_image in zip(targets, matched_idxs):
+            gt_classes_targets = targets_per_image["labels"][matched_idxs_per_image.clip(min=0)]
+            gt_classes_targets[matched_idxs_per_image < 0] = -1  # backgroud
+            gt_boxes_targets = targets_per_image["boxes"][matched_idxs_per_image.clip(min=0)]
+            all_gt_classes_targets.append(gt_classes_targets)
+            all_gt_boxes_targets.append(gt_boxes_targets)
+
+        all_gt_classes_targets = torch.stack(all_gt_classes_targets)
+        # compute foregroud
+        foregroud_mask = all_gt_classes_targets >= 0
+        num_foreground = foregroud_mask.sum().item()
+
+        # classification loss
+        gt_classes_targets = torch.zeros_like(cls_logits)
+        gt_classes_targets[foregroud_mask, all_gt_classes_targets[foregroud_mask]] = 1.0
+        loss_cls = sigmoid_focal_loss(cls_logits, gt_classes_targets, reduction="sum")
+
+        # regression loss: GIoU loss
+        pred_boxes = [
+            self.box_coder.decode_single(bbox_regression_per_image, anchors_per_image)
+            for anchors_per_image, bbox_regression_per_image in zip(anchors, bbox_regression)
+        ]
+        # amp issue: pred_boxes need to convert float
+        loss_bbox_reg = giou_loss(
```

Comment:
Can't we use the already existing[ generalised box iou ](https://github.com/pytorch/vision/blob/main/torchvision/ops/boxes.py) here (`1 - generalized_box_iou`)?

Reply:
there is difference bettween `generalised box iou` and `giou_loss`. The `generalised box iou` compute the giou between each pair of elements of two set boxes, the shape output is [N, M]. If we use `generalised box iou` to get giou loss, you must  extract diagonal element, as we only want to get one-2-one giou.

Reply:
I was doing some experiments and seems like `1 - generalized_box_iou(tensor1, tensor2).diagonal()` is equivalent, but because it does all to all (as you said) is considerably slower. In that case I think we can keep both implementations but I suggest to move the giou_loss to `ops.boxes` alongside `generalized_box_iou` and perhaps renaming to `generalized_box_iou_121`

Reply:
> I was doing some experiments and seems like `1 - generalized_box_iou(tensor1, tensor2).diagonal()` is equivalent, but because it does all to all (as you said) is considerably slower. In that case I think we can keep both implementations but I suggest to move the giou_loss to `ops.boxes` alongside `generalized_box_iou` and perhaps renaming to `generalized_box_iou_121`

maybe `generalized_box_iou_loss` is better?

Reply:
sounds good

Reply:
@jdsgomes I found that `focal_loss` is a single file:
https://github.com/pytorch/vision/blob/main/torchvision/ops/focal_loss.py
so I think `generalized_box_iou_loss`  should also. Maybe all loss function can be placed in a `loss` directory or `losses.py`.

Reply:
that would be backwards incompatible, so if we ever decide to do that we need to coordinate and make the changes in separate PRs.

For now I would then leave it standalone just update the name.

Reply:
@jdsgomes OK, you can update the name.